Disallow partial results when shard unavailable

henningandersen opened the pull request:
Searching with `allowPartialSearchResults=false` could still return
partial search results during recovery. If the last shard copy fails
with a "shard not available" exception, the failure would be ignored and
a partial result returned. The one case where this is known to happen
is when the last shard copy is recovering when searching, since
`IllegalIndexShardStateException` is considered a "shard not available"
exception.

Relates to #42612

I wonder if this should go back to 7.3.1 and 6.8.x too?

henningandersen committed:
Disallow partial results when shard unavailable

Searching with `allowPartialSearchResults=false` could still return
partial search results during recovery. If the last shard copy fails
with a "shard not available" exception, the failure would be ignored and
a partial result returned. The one case where this is known to happen
is when the last shard copy is recovering when searching, since
`IllegalIndexShardStateException` is considered a "shard not available"
exception.

Relates to #42612

elasticmachine commented:
Pinging @elastic/es-search

jimczi reviewed (commented):
While it would be nice to fail requests early in case of shard failures I wonder why this pr is marked as a bug. `AbstractSearchAsyncAction#onPhaseDone` should check shard failures and raises an error if some shards failed and allowPartialResults is set to false. Did I miss something ?

jimczi reviewed (commented):
Sorry I didn't check AbstractSearchAsyncAction#onShardFailure which ignores shard not available exceptions. This does sound like a bug ;). However I am not sure we should raise a phase failure while other requests are in flight, it sounds less risky to do the verification in `onPhaseDone` like we do today but we should compare the total number of shards with `successfulOps + skippedOps` in order to determine if a partial failure occurred or not ?

henningandersen committed:
Fix checkstyle.

henningandersen committed:
Redo check for partial failure

Check for partial failure by comparing successful == num-shards to avoid
bailing out early in the search phase.

Fixed race condition in MockSearchService.

henningandersen committed:
Revert unnecessary change.

henningandersen committed:
Rename test.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_search_allow_partial_results_false

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_search_allow_partial_results_false

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_search_allow_partial_results_false

henningandersen commented:
Thanks @jimczi , this indeed seems less risky. Please have another look at your convenience.

jimczi commented on server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -156,6 +156,16 @@ public final void executeNextPhase(SearchPhase currentPhase, SearchPhase nextPha
                     return;
                 }
             }
+            if (allowPartialResults == false && successfulOps.get() != getNumShards()) {
```
We shouldn't have two tests to check if a partial failure occurred. This one could replace the block above and if successfulOps is less than the number of shards and shardSearchFailures.length is 0 we can assume that the failure was a shard unavailable exception ?

henningandersen replied:
Fixed in c2df1ab6996e2e35f5e7fc8bbf139fdadca09c93

jimczi commented on server/src/test/java/org/elasticsearch/search/basic/SearchRedStateIndexIT.java:
```diff
@@ -95,6 +99,89 @@ public void testClusterDisallowPartialsWithRedState() throws Exception {
     }
 
+    public void testDisallowPartialsWithRedStateRecovering() throws Exception {
+        int docCount = scaledRandomIntBetween(1000, 10000);
```
do we need this number of docs for the test ? `10, 100` should be enough, no ?

henningandersen replied:
The test has a tendency to fail more often on larger sizes. I ended up with 10, 1000, since this makes it fail around 1/3 of the runs (but this is likely highly dependent on the hardware). Changed in: c2df1ab6996e2e35f5e7fc8bbf139fdadca09c93

jimczi reviewed (commented):
Thanks for updating @henningandersen , I left more comments

henningandersen committed:
Improve code and reduce indexing

Index less docs to reduce runtime of test and improve code checking for
partial shard failures to only check flag/successful ops once.

henningandersen commented:
Failures in bwc and default-distro look similar to #45605, both times failing on 7.4 mixed cluster tests. Retesting:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_search_allow_partial_results_false

henningandersen commented:
ci/2 failed with this new issue: 45866, hope it works on a retry:
@elasticmachine run elasticsearch-ci/2

henningandersen commented:
Thanks for reviewing @jimczi , this is ready for another round.

jimczi commented on server/src/test/java/org/elasticsearch/search/basic/SearchRedStateIndexIT.java:
```diff
@@ -95,6 +99,89 @@ public void testClusterDisallowPartialsWithRedState() throws Exception {
         assertThat(ex.getDetailedMessage(), containsString("Search rejected due to missing shard"));        
     }
 
+    public void testDisallowPartialsWithRedStateRecovering() throws Exception {
```
I wonder if it would be simpler to add a unit test in `FetchSearchPhaseTests` and `AbstractSearchAsyncActionTests`? We just need a child query that throws a ShardNotAvailableException and the reason why the exception is thrown is not really important. I also don't get why you need a test that fails reasonably often ? Shouldn't it fail all the time ?

jimczi reviewed (commented):
I left a comment regarding the IT test. Thanks for updating the partial failure condition.

henningandersen committed:
Removed integration test

To be introduced separately as a disruption test.

henningandersen committed:
Revert a little more.

henningandersen committed:
Reverted whitespace changes

henningandersen commented:
Spoke to @jimczi about this on another channel. We decided to pull out the integration test into a separate PR as a starting point for a disruptive style search tests.

@jimczi, this is ready for another round.

jimczi reviewed (approved):
LGTM

henningandersen commented:
ci/1 seems to be caused by the `ShardLockObtainFailedException` failure reported in #45605 .
@elasticmachine run elasticsearch-ci/1

henningandersen commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix_search_allow_partial_results_false
